### DISCUSSION TITLE
Dibs

### DISCUSSION THREAD

**Discussion by zeddie888** [UWCubeSat/cpp-boot-camp#2] [General]
Edward
- [x] Pointers
- [x] Unique pointers
- [x] Shared pointers
- [x] Copy constructor
- [ ] Copy-assignment
- [ ] Virtual/non-virtual methods
- [ ] Iterators

Tri
- [X] Binary/Hexdecimal
- [x] Primitive Types
- [x] Structs/Enums/etc.
- [x] Memory management
- [ ] Compiling
    - [ ] Header Files
- [ ] Functions
   - [ ] Operator Overload
   - [ ] Lambda Functions
- [ ] STL Library

**Comment by markasoftware** [comment]
Also should cover
+ [ ] Virtual vs non-virtual methods
+ [ ] References (and differences from pointers)
+ [ ] Understanding `hpp` vs `cpp` files (maybe this is included in "compiling"?)
+ [x] Copy constructors
+ [ ] Operator overloading
+ [ ] RAII (probably included in "memory management"). It's just as important for people to know when things will get automatically deallocated as to know how to manually allocate memory.
+ [ ] Iterators, including how pointers are technically iterators.

(Basically anything commonly used that's different from java)

**Comment by zeddie888** [comment]
Unclaimed:

- [ ] STL
    - [ ] Vectors
    - [ ] Data Structures (Maps, Sets, Trees, etc.)
    - [ ] Casting Operators
- [ ] Functions
    - [ ] Templates
    - [ ] Lambda expressions
- [ ] Variables
    - [x] Qualifiers (`const`, `extern`, etc.)
    - [ ] Move semantics
- [ ] Classes
    - [ ] Constructors
    - [ ] Destructors
    - [ ] Qualifiers (`const`, `virtual`, etc.)
    - [ ] Inheritance
    - [ ] Operator Overloading
    - [ ] Templates
    - [ ] Move Semantics
- [ ] Casting
- [ ] Make a reference back to the table of Contents at the End

**Comment by nguy8tri** [comment]
Let's build an overall structure to this:
1. Basics of C++
    - Hex/Binary
    - STL
3. Variables
    - Primitives
    - C/C++ Unique
4. Functions (Need to think more)
5. Pointers
    - Smart Pointers
    - Iterators
    - Function Pointers
6. Memory
    - Compilation
    - Memory Format
7. Classes (Lots to include)

This structure is a bit off. There are some circular sections (i.e. Section1 requires Section2, which requires Section1 -> Think of graphs and cycles) with how we've written these sections, or maybe its just me. We can just keep on writing, and restructure afterwards.
